ImGuiFileDialog.cpp

```diff
@@ -75,300 +103,330 @@ ImGuiFileDialog::~ImGuiFileDialog()
 }
 
 /* Alphabetical sorting */
-static int alphaSort(
-	const void *a, const void *b)
+static int alphaSort(const struct dirent **a, const struct dirent **b)
 {
-	return strcoll(((dirent*)a)->d_name, ((dirent*)b)->d_name);
+    return strcoll((*a)->d_name, (*b)->d_name);
 }
 
-static bool stringComparator(FileInfoStruct a, FileInfoStruct b)
-{
-	bool res;
-	if (a.type != b.type) res = (a.type < b.type);
-	else res = (a.fileName < b.fileName);
-	return res;
-}
 
 void ImGuiFileDialog::ScanDir(std::string vPath)
 {
-	struct dirent **files = 0;
-	int i = 0;
-	int n = 0;
-
-	m_FileList.clear();
-
-	if (m_CurrentPath_Decomposition.size() == 0)
-	{
-		// get currentPath
-		DIR *currentDir = opendir(vPath.c_str());
-		if (currentDir == 0) // path not existing
-		{
-			vPath = "."; // current  app path
-			currentDir = opendir(vPath.c_str());
-		}
-		if (currentDir != 0)
-		{
-			std::wstring ws(currentDir->wdirp->patt);
-			m_CurrentPath = std::string(ws.begin(), ws.end());
-			ReplaceString(m_CurrentPath, "\\*", "");
-			closedir(currentDir);
-			m_CurrentPath_Decomposition = splitStringVector(m_CurrentPath, '\\');
-		}
-		else
-		{
-			return;
-		}
-	}
-	
-	/* Scan files in directory */
-	n = scandir(vPath.c_str(), &files, NULL, alphaSort);
-	if (n >= 0)
-	{
-		for (i = 0; i < n; i++)
-		{
-			struct dirent *ent = files[i];
-
-			FileInfoStruct infos;
-
-			infos.fileName = ent->d_name;
-
-			if (infos.type == 'f')
-			{
-				size_t lpt = infos.fileName.find_last_of(".");
-				if (lpt != std::string::npos)
-					infos.ext = infos.fileName.substr(lpt);
-			}
-			
-			if (infos.fileName != ".")
-			{
-				switch (ent->d_type)
-				{
-				case DT_REG: infos.type = 'f'; break;
-				case DT_DIR: infos.type = 'd'; break;
-				case DT_LNK: infos.type = 'l'; break;
-				}
-
-				m_FileList.push_back(infos);
-			}
-		}
-
-		for (i = 0; i < n; i++)
-		{
-			free(files[i]);
-		}
-		free(files);
-	}
-
-	std::sort(m_FileList.begin(), m_FileList.end(), stringComparator);
+    struct dirent **files               = NULL;
+    int             i                   = 0;
+    int             n                   = 0;
+
+    m_FileList.clear();
+
+    if (0u == m_CurrentPath_Decomposition.size())
+    {
+        SetCurrentDir(vPath);
+    }
+
+    if (0u != m_CurrentPath_Decomposition.size())
+    {
+        n = scandir(vPath.c_str(), &files, NULL, alphaSort);
+        if (n > 0)
+        {
+            for (i = 0; i < n; i++)
+            {
+                struct dirent *ent = files[i];
+
+                FileInfoStruct infos;
+
+                infos.fileName = ent->d_name;
+                if (("." != infos.fileName) && (".." != infos.fileName))
```

Comment:
I removed the ".." path too, but then later saw you had special handling for it. I'm not quite sure how you were using it originally, but with the separated path at the top of the widget it isn't really needed.